[browser] Add tabs for mode private. Contributes to JB#50860

xlise opened the pull request.

llewelld commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +32,144 @@ SilicaGridView {
-                if (remorsePopup) {
-                    remorsePopup.trigger()
-                }
+    PrivateModeTexture {
```
Currently the private mode background texture appears/disappears across the whole page as the tab is changed. I'm wondering if maybe it would look better if it slid in along with the page?

xlise replied:
I agree, corrected

llewelld commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +32,144 @@ SilicaGridView {
+                id: _persistentTabView
+                privateMode: false
+                portrait: tabs.portrait
+                model: !tabs.privateMode ? tabs.tabModel : null
```
Using the same `tabModel` and switching between them like this seems to be causing problems for the animation when swiping between the normal and private tab listings. The items appear suddenly half way through the swipe.

To get this to work nicely, I think it'll be necessary to maintain both models simultaneously and separately here.

xlise replied:
I agree, corrected

llewelld commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -11,26 +11,18 @@
-SilicaGridView {
-    id: tabView
+Private.TabView {
+    id: tabs
```
There's an error from Overlay.qml attempting to access `Browser.TabView.positionViewAtIndex()` It doesn't look like it's being caused by your changes here, but do you think it would be possible to fix it along with these changes?
```
[W] unknown:656 - file:///usr/share/sailfish-browser/pages/components/Overlay.qml:656: ReferenceError: positionViewAtIndex is not defined
[W] unknown:665 - file:///usr/share/sailfish-browser/pages/components/Overlay.qml:665: TypeError: Property 'positionViewAtIndex' of object TabView_QMLTYPE_144
```

xlise replied:
Applied the hack, as there are no ideas how to do it correctly 
```
        Component.onCompleted:  {
            if ((index / columns * cellHeight) > tabView.height) {
                tabView.loaded = true
            }
        }
```

xlise commented on apps/browser/qml/pages/components/TabGridView.qml:
```diff
@@ -0,0 +1,141 @@
+        defaultValue: true
+    }
+
+    PullDownMenu {
```
Return to the previous page blocked, the menu is still left

adenexter replied:
The gesture conflict could be reduced by disabling wrapping in the tab view. Then if you're on the first tab the gesture would return you to the previous page.

`wrapMode: PagedView.NoWrap` would do I think.

xlise replied:
Good idea, I will consult with the designer

xlise replied:
Fixed
`interactive: false` It turned out like this

xlise commented on apps/browser/qml/pages/components/TabGridView.qml:
```diff
@@ -0,0 +1,141 @@
+
+    width: parent.width - Theme.horizontalPageMargin
+    height: parent.height
+    x: Theme.horizontalPageMargin
```
It turned out to be problematic to fix, besides, this problem will go away when switching to the popup menu

adenexter replied:
The convention is to put the menu above the header. The tab header background rectangle is going to create a bit of a problem there because it would need to be below the pull down menu so it doesn't occlude the highlight.

But for reference the way to position the menu at the top of the window would be to assign the tabs grid view to the TabItem flickable property and add a header item to the grid view that is the same height as TabView header.

```
Private.TabItem {
    flickable: _persistentTabView
    TabGridView {
         id: _persistentTabView
         header: Item {
              width: 1
              height: tabs.headerHeight
         }
   }
}
```

The horizontal offset might still be a problem even then though. Provided there's no clipping though, you might be able to fix it by adding a negative x to the PullDownMenu as a counter margin and increasing the width of the menu to resolve the margin on the right.

Alternatively including the margin in the GridView delegates instead of the view itself would work too.

xlise replied:
Thanks for your suggestions for improvement. Fixed gap. Moved the pulley menu above the title, but this can only be considered as a temporary solution before the implementation of the popup menu

xlise commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +32,144 @@ SilicaGridView {
+    Component.onCompleted: {
+        persistentIcon.updateGrubImage()
+        privateIcon.updateGrubImage()
+        currentIndex = privateMode
```
done

xlise commented:
![Снимок_Экрана_20210520_001](https://user-images.githubusercontent.com/5721108/118961646-34013400-b96d-11eb-92db-f7039f9fe941.png)

xlise commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +34,204 @@ SilicaGridView {
+                    source: webView.privateTabModel.count > 0 ? "image://theme/icon-m-incognito-selected" : "image://theme/icon-m-incognito"
+                    visible: false
+                }
+                Rectangle {
```
Yes you are right. done

adenexter commented:
It's looking pretty good to me now. I am seeing a single pixel uneven line just below the pulley menu indicator. I think this is the normal tab header background item not being quite hidden by the black rectangle background and on just that one line you can see through to the web page below the page.

Hiding the tab header background with `_headerBackgroundVisible: false` in the silica TabView fixed that for me.

The scroll bar also gets hidden behind the tab header when the view is scrolled to the top but I don't have a solution for that, it is within a tab while needing to appear above a sibling item to the tab in the header.

llewelld commented on apps/browser/qml/pages/components/TabGridView.qml:
```diff
@@ -0,0 +1,153 @@
+
+    width: parent.width - Theme.horizontalPageMargin
+    height: parent.height
+    y: Theme.paddingLarge
```
Shifting the `SilicaGridView` down by `Theme.paddingLarge` here is unfortunately causing the pulley menu to be shifted down as well, so there's a gap at the top when you pull the items down.

I recommend removing this `y` shift here, and instead adding `Theme.paddingLarge` to both `TabGridView.header` items in `TabView.qml`.
```suggestion
```

xlise replied:
done

llewelld commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +34,230 @@ SilicaGridView {
+                    model: webView.privateTabModel
+                    header: Item {
+                        width: 1
+                        height: tabs.tabBarHeight
```
Similarly, the same extra space would need to be added here:
```suggestion
                        height: tabs.tabBarHeight + Theme.paddingLarge
```

xlise replied:
done

llewelld commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +34,230 @@ SilicaGridView {
+                    model: webView.persistentTabModel
+                    header: Item {
+                        width: 1
+                        height: tabs.tabBarHeight
```
In order to compensate for removing the y offset from `TabGridView.qml` I suggest adding the extra space here instead.
```suggestion
                        height: tabs.tabBarHeight + Theme.paddingLarge
```

xlise replied:
done

llewelld reviewed (changes requested):
This is so much nicer, thank you. You even managed to solve the "swipe back" problem, which is fantastic. I've made just a few small comments, mostly about fixing some spacing issues. There also seem to be a couple of comments from Andrew still unresolved if I'm reading correctly.

xlise commented:
> This is so much nicer, thank you. You even managed to solve the "swipe back" problem, which is fantastic. I've made just a few small comments, mostly about fixing some spacing issues. There also seem to be a couple of comments from Andrew still unresolved if I'm reading correctly.

Notes from Andrew will still edit

xlise commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +34,230 @@ SilicaGridView {
+            persistentIcon.updateGrubImage()
+            privateIcon.updateGrubImage()
+            currentIndex = privateMode ? 1 : 0
         }
```
Thank you, done

xlise commented on apps/browser/qml/pages/components/TabGridView.qml:
```diff
@@ -0,0 +1,153 @@
+    y: Theme.paddingLarge
+    x: Theme.horizontalPageMargin
+    currentIndex: -1
+    header: spacer
```
Thank you, it's really neat. done

xlise commented on apps/browser/qml/pages/components/TabGridView.qml:
```diff
@@ -0,0 +1,153 @@
+        id: spacer
+        Item {
+            width: tabView.width
+            height: Theme.paddingMedium
```
done

llewelld commented:
> The scroll bar also gets hidden behind the tab header when the view is scrolled to the top but I don't have a solution for that, it is within a tab while needing to appear above a sibling item to the tab in the header.

Maybe this is really nasty, but placing the `VerticalScrollBarDecorator` inside an `Item` that's only as high as the area beneath the toolbar could work. Something like this (this would need tweaking though because `tabsToolBar.height` isn't currently available inside the `TabViewGrid`s).

```
Item {
    y: tabsToolBar.height
    height: parent.height - y
    anchors.right: parent.right
    width: Math.round(Theme.paddingSmall/2)

    VerticalScrollDecorator {
        _forcedParent: parent
        flickable: tabView
    }
}
```

llewelld commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +34,230 @@ SilicaGridView {
+            Rectangle {
+                anchors {
+                    fill: parent
+                    topMargin: _pulleyMenuApparentHeight
```
I really like the way you've handled hiding the top menu highlight; it's very ingenious. However it seems to cause some issues, because `_pulleyMenuApparentHeight` doesn't update when the normal/private tab changes. I tried calling `_changeMenuApparentHeight()` when the `currentIndex` changed, but that resulted in other animation artefacts.

Perhaps as an alternative consider adding in a covering bar to the two `Private.TabItem` items instead. This way the bar will slide left/right when the grids slide. Something like this (just a rough idea):
```
Rectangle {
    color: Theme.colorScheme == Theme.LightOnDark ? "black" : "white"
    width: parent.width
    height: Math.max(0, _pulleyMenuHeight - y)
    y: Math.max(0, _pulleyMenuHeight - _persistentTabView.contentY + _persistentTabView.originY)
}   
```
It makes the code a bit messy, but then this is only a temporary solution until the menu is changed anyway.

xlise replied:
Thank you, so much better. done

llewelld replied:
Thanks @xlise. Unfortunately I don't see the change in practice (did you push it to the repo? I see a force-push, but I don't see any changes; it looks like a rebase).

xlise replied:
Sorry, I haven't added any files. done

llewelld replied:
Thank you! You should be safe to remove the `_changeMenuApparentHeight()` logic now, and replace the header topMargin to be a fixed height. I'll add a comment about it above.

llewelld commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +34,248 @@ SilicaGridView {
+            Rectangle {
+                anchors {
+                    fill: parent
+                    topMargin: _pulleyMenuApparentHeight
```
This should now be safe:
```suggestion
                    topMargin: _pulleyMenuHeight
```
Which would allow you to remove the `_changeMenuApparentHeight()` logic I think.

xlise replied:
I agree, removed

llewelld reviewed (approved):
Thanks for your work on this, it looks good to me, and is working really great in my opinion.

Please rebase before we merge it.

xlise commented:
> Thanks for your work on this, it looks good to me, and is working really great in my opinion.
> 
> Please rebase before we merge it.

Yes, it works, but I still haven't corrected one comment from Andrew, but it doesn't work there as Andrew suggested

llewelld commented:
Thanks, and understood. I won't merge anything until you've completed your discussion with Andrew and you say it's ready (it still needs approval from someone at OMP too before we can merge it).

xlise commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +34,204 @@ SilicaGridView {
+
+        function _updateHeaderModel() {
+            headerModel.set(0, { "icon": persistentIcon.grabIcon })
+            headerModel.set(1, { "icon": privateIcon.grabIcon })
```
Thank you, this option is really better. done

xlise commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +34,204 @@ SilicaGridView {
+                id: privateIcon
+                property string grabIcon
+
+                function updateGrubImage() {
```
done

adenexter reviewed (commented):
> Yes, it works, but I still haven't corrected one comment from Andrew, but it doesn't work there as Andrew suggested. Rebased

Which comment was this? The only thing I can see that might be unresolved is that you could use a delegate and model for the Silica TabView the same as with the header so as to not have two mostly the same TabItem components.

adenexter commented on apps/browser/qml/pages/components/TabView.qml:
```diff
@@ -40,131 +34,221 @@ SilicaGridView {
+                z: -100
+                color: Theme.colorScheme == Theme.LightOnDark ? "black" : "white"
             }
+            Component.onCompleted: tabs.headerHeight = Qt.binding(function() { return height })
```
Correction, Silica TabView has a tabBarHeight property so you can use that instead of defining your own headerHeight property.

xlise replied:
Thanks!  done

xlise commented:
> > Yes, it works, but I still haven't corrected one comment from Andrew, but it doesn't work there as Andrew suggested. Rebased
> 
> Which comment was this? The only thing I can see that might be unresolved is that you could use a delegate and model for the Silica TabView the same as with the header so as to not have two mostly the same TabItem components.

Yes, I mean this remark, I corrected the TabBar. delegate for components today I will try to make

adenexter commented:
> Yes, I mean this remark, I corrected the TabBar. delegate for components today I will try to make

Okay. Something to watch for is TabGridView has both model and privateMode properties, so if you try and reference the model data from within the TabGridView declaration both model.privateMode and privateMode will resolve to its properties and not the model data.

The best way to address that is probably bind model.privateMode to a property in the delegate top level TabItem and reference that.

xlise commented:
> > Yes, I mean this remark, I corrected the TabBar. delegate for components today I will try to make
> 
> Okay. Something to watch for is TabGridView has both model and privateMode properties, so if you try and reference the model data from within the TabGridView declaration both model.privateMode and privateMode will resolve to its properties and not the model data.
> 
> The best way to address that is probably bind model.privateMode to a property in the delegate top level TabItem and reference that.

Great thanks a lot for your hint. The code is much better. done

aroditelev commented:
LGTM
